### ISSUE TITLE
Custom error statusCode

### ISSUE THREAD

**Issue by sebdeckers** [fastify/fastify#401]
(Apologies for the simple question. I'm completely new to Fastify and could not yet find the answer in the docs/examples/issues/source.)

## Question

Can a thrown Error instance set the outgoing HTTP status code?

## Use Cases

- Output specific status codes with convenient error throwing from async functions.
- https://www.npmjs.com/package/http-errors

## Example

```js
const fastify = require('fastify')({ logger: true })

fastify.get('/', async function (request, reply) {
  const error = new Error()
  error.message = 'This is the message'
  error.statusCode = 404
  error.error = 'This is the error'
  throw error
})

fastify.listen(3000, function (err) {
  if (err) throw err
  console.log(`server listening on ${fastify.server.address().port}`)
})
```

## Actual Output

```
$ curl --verbose http://[::1]:3000/
*   Trying ::1...
* TCP_NODELAY set
* Connected to ::1 (::1) port 3000 (#0)
> GET / HTTP/1.1
> Host: [::1]:3000
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 500 Internal Server Error
< Content-Type: application/json
< Content-Length: 82
< Date: Mon, 30 Oct 2017 04:17:12 GMT
< Connection: keep-alive
< 
* Connection #0 to host ::1 left intact
{"error":"Internal Server Error","message":"This is the message","statusCode":500}
```

## Expected Output

```
< HTTP/1.1 404 Internal Server Error
[...]
{"error":"Internal Server Error","message":"This is the message","statusCode":404}
```

**Comment by delvedor** [comment]
Hi! Thank you for checking out Fastify!
You can do that in three ways:
```js
// With standard function
fastify.get('/', (req, reply) => {
   reply.code(404).send(new Error('Missing this'))
})

// With async  function
fastify.get('/', async (req, reply) => {
   reply.code(404)
   return new Error('Missing this')
})

// Using throw inside an async  function
fastify.get('/', async (req, reply) => {
   reply.code(404)
   throw new Error('Missing this')
})
```

**Comment by mcollina** [comment]
I thought that was supported.
IMHO we should change this block (https://github.com/fastify/fastify/blob/master/lib/reply.js#L176-L203) so that it checks if an error has a `statusCode`  property, and use that instead of 500.
Would you mind opening up a PR?

**Comment by dound** [comment]
@mcollina @sebdeckers  This behavior surprised me!

I had a bug in my code. My code was using the AWS SDK and failed to catch one of the exceptions it could throw. I was confused because when that exception was thrown (and uncaught) my API response code was 400 -- I had expected an uncaught exception to trigger a 500 error. After some digging, I realized the uncaught AWS exception happened to have a field named "statusCode" and that fastify was using that.

I've since fixed the bug in my code which left this third-party exception unhandled. But in case I miss other third-party exceptions which unwittingly trigger this special behavior in fastify, I've updated my custom error handler to check for the statusCode field in the uncaught exception, and if present set the status code to 500 (i.e., uncaught exceptions should always be a 500 error in my codebase).

Perhaps this behavior should be documented on [Errors](https://www.fastify.io/docs/v3.1.x/Errors/) page? Apologies if it is documented somewhere else and I overlooked it.
